fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/FindPossibleIndexForAndComponentRule.java

```diff
@@ -0,0 +1,119 @@
+ * and produces a {@link RelationalGroupRefHolder} containing a plan that scans each of those indexes. These can then
+ * be planned individually and then compared using the {@link PickFromPossibilitiesRule}.
+ *
+ * <p>
+ * Currently, this rule is pretty rough. Among other things, it does not have proper support for using the primary key
+ * index and it pushes the entire filter into each of the members of the
+ * {@link com.apple.foundationdb.record.query.plan.temp.FixedCollectionExpressionRef}, rather than removing the
+ * comparison that is already planned. This rule will probably change substantially as the planner improves.
+ * </p>
+ */
+@API(API.Status.EXPERIMENTAL)
+public class FindPossibleIndexForAndComponentRule extends PlannerRule<LogicalFilterExpression> {
+    private static ExpressionMatcher<FieldWithComparison> fieldMatcher = TypeMatcher.of(FieldWithComparison.class);
+    private static ExpressionMatcher<AndComponent> andFilterMatcher = TypeMatcher.of(AndComponent.class,
+            AllChildrenMatcher.allMatching(fieldMatcher));
+    private static ExpressionMatcher<RecordQueryScanPlan> scanMatcher = TypeMatcher.of(RecordQueryScanPlan.class);
+    private static ExpressionMatcher<LogicalFilterExpression> root = TypeMatcher.of(LogicalFilterExpression.class,
+            andFilterMatcher, scanMatcher);
+
+    public FindPossibleIndexForAndComponentRule() {
+        super(root);
+    }
+
+    @Nonnull
+    @Override
+    public ChangesMade onMatch(@Nonnull PlannerRuleCall call) {
+        if (!call.get(scanMatcher).hasFullRecordScan()) {
+            return ChangesMade.NO_CHANGE; // already has some bounds on the scan
+        }
+
+        List<FieldWithComparison> fields = call.getBindings().getAll(fieldMatcher);
+        List<RelationalPlannerExpression> possibilities = new ArrayList<>();
+        KeyExpressionComparisons keyComparisons;
+        Optional<KeyExpressionComparisons> matchedKeyComparisons;
+
+        KeyExpression commonPrimaryKey = call.getContext().getCommonPrimaryKey();
+        if (commonPrimaryKey != null) {
+            keyComparisons = new KeyExpressionComparisons(commonPrimaryKey);
+            for (FieldWithComparison field : fields) {
+                matchedKeyComparisons = keyComparisons.matchWith(field);
+                if (matchedKeyComparisons.isPresent()) {
+                    // TODO improve with a logical version of a RecordQueryScanPlan that holds a KeyExpressionComparisons.
+                    possibilities.add(new LogicalFilterExpression(Query.and(fields),
+                            new RecordQueryScanPlan(matchedKeyComparisons.get().toScanComparisons(), false)));
+                    break;
+                }
+            }
+        }
+
+        for (Index index : call.getContext().getIndexes()) {
+            keyComparisons = new KeyExpressionComparisons(index.getRootExpression());
```

Comment:
I believe there had been some discussion of putting the primary key into the `KeyExpressionComparisons` here. Is that one of the things that makes this class "rough" (according to the Javadoc)?